File: Java/src/Solution001_100/Solution20.java
package Solution001_100;

import java.util.HashMap;
import java.util.Stack;

/**
 * @author windows
 */
public class Solution20 {

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack();
        HashMap<Character, Character> map = new HashMap<>();
        map.put('}', '{');
        map.put(']', '[');
        map.put(')', '(');
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '{' || s.charAt(i) == '[' || s.charAt(i) == '('){
                stack.push(s.charAt(i));
            }else {
                if (stack.isEmpty()) {
                    return false;
                }else {
                    Character character = map.get(s.charAt(i));
                    if (!character.equals(stack.pop())) {
                        return false;
                    }
                }
            }
        }
        if (stack.isEmpty()){
            return true;
        }else{
            return false;
        }
    }

    public static void main(String[] args) {
        Solution20 solution20 = new Solution20();
        boolean valid = solution20.isValid("[[[[");
        System.out.println(valid);
    }
}

File: C++/src/Solution1_100/Solution20.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution20.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <stack>
#include <unordered_map>

bool Solution20::isValid(string s) {
    stack<char> stack;
    unordered_map<char, char> map = {{'}', '{'}, {')', '('}, {']', '['}};
    for (char c : s) {
        if (!map.count(c)){
            stack.push(c);
        }else {
            if (stack.empty() || stack.top() != map[c]){
                return false;
            }
            stack.pop();
        }
    }
    return stack.empty();
}
